Fix: Never completionPolicy ImagePullJob timeout setting should consider backoffLimit

MajLuu opened the pull request:
issue: [2071](https://github.com/openkruise/kruise/issues/2071)

ABNER-1 commented:
/lgtm  
@zmberg, could you please review this?

MajLuu commented:
@zmberg @FillZpp @veophi , could you please review this?

MajLuu commented:
> @MajLuu Currently timeout is pretty clear and we use it heavily internally. Is there an actual scenario on your end that the current logic can't accommodate?

In our machine learning cluster, we set the pull timeout for a single image to 1 hour, which can be retried 3 times; at the same time, we started many never completionPolicy image pull jobs (10+). It is estimated that the size of our machine learning images (including cuda, pytorch, etc.) is more than 20GB. Most nodes can pull images down after running normally for a period of time, but the newly added nodes in the cluster suddenly received 10+ image pull jobs, which resulted in 1 hour of failure to download the images. At this time, setting the timeout to 3*1h can solve this scenario. In this scenario, we do not think it is reasonable to change the timeout to 3h * 1.

furykerry commented:
@MajLuu we've added command line argument for controlling concurrent imagepulling workers, maybe setting the worker limits can help increasing the effective image pulling efficiency for newly added nodes. plz check the patch for more detail: https://github.com/openkruise/kruise/commit/318165b7ea75226b71c704c332c92de9fa9dc274

MajLuu commented:
> @MajLuu we've added command line argument for controlling concurrent imagepulling workers, maybe setting the worker limits can help increasing the effective image pulling efficiency for newly added nodes. plz check the patch for more detail: [318165b](https://github.com/openkruise/kruise/commit/318165b7ea75226b71c704c332c92de9fa9dc274)

OK, we have also implemented the waiting mechanism on the internal 1.4 version, and configured it to pull a maximum of 3 images. At the same time, Containerd has implemented a mechanism to limit the number of concurrent pulling images. However, there is still a problem of excessive disk IO, so this codes was modified. In our machine learning clusters, users can wait for the image to be downloaded; however, there is a strict container startup timeout limit in our flink clusters, and the container image must be guaranteed to exist on the node (there are a large number of pods in the flink nodes, which leads to a large IO utilization rate of the node). The image pull job created is never pullPolicy. If the image cannot be downloaded normally within the specified time, it can only be continued after 24 hours. The problem mentioned in the issue was discovered when troubleshooting the problem of inaccurate download timeouts. If the community considers this modification is unnecessary, please close the issue. Thank you~
